src/main/java/org/spongepowered/api/item/inventory/menu/SlotClickHandler.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.inventory.menu;
+
+import org.spongepowered.api.item.inventory.Container;
+import org.spongepowered.api.item.inventory.Slot;
+import org.spongepowered.api.item.inventory.slot.SlotIndex;
+
+import javax.annotation.Nullable;
+
+@FunctionalInterface
+public interface SlotClickHandler {
+
+    /**
+     * Handler for inventory click callbacks.
+     *
+     * @param container the container
+     * @param slot the primary clicked slot
+     * @param slotIndex the primary clicked slot index
+     * @param clickType the type of click
+     * @param slot2 the secondary slot in the interaction. not null for {@link ClickTypes#KEY_SWAP}
+     *
+     * @return false to prevent the click
+     */
+    boolean handle(Container container, Slot slot, SlotIndex slotIndex, ClickType clickType, @Nullable Slot slot2);
```

Comment:
I don't like the second `slot` as a parameter, maybe we just need specialized click handlers in those cases. IMO there are too many arguments and adding new ones for every special case isn't good.

The SlotIndex is based on the `Container`? Then the parameter is unnecessary unless for convenience.

```java
interface SlotClickHandler {
    boolean handle(Container container, ClickType clickType, Slot slot);

    default boolean handleKeySwap(Container container, Slot hoveredSlot, int key) {
        return handle(container, ClickTypes.KEY_SWAP, hoveredSlot);
    }
}
```

Another thought, what happens when you return `true`/`false`? Does this prevent default behavior if you return `false`?

Maybe it's even an option to allow different `ClickType`s to use different handler interfaces.

Reply:
I guess I could make a second handler interface for the KEY_SWAP type
`SlotIndex` is indeed for convenience.
As the javadocs say. You return false to prevent the click from happening.

Reply:
Now there is
ClickHandler - no slot
SlotClickHandler - one slot
KeySwapHandler - two slots
also
CloseHandler
SlotChangeHandler

They can all be registered with one class implementing all of them. `#registerHandler`